File: notebooks/add_videos.js
/**
 * Script pour ajouter des cellules d'enregistrement vidéo aux notebooks.
 * Exécute: node add_videos.js
 */
const fs = require('fs');
const path = require('path');

function addVideoCell(notebookPath, config) {
    const { envName, nEpisodes, agentType, modelPath, videoPrefix, description } = config;
    
    console.log(`\n📝 Traitement de ${notebookPath}...`);
    
    const nb = JSON.parse(fs.readFileSync(notebookPath, 'utf8'));
    
    // Construire le code selon le type d'agent
    let code;
    if (agentType === 'random') {
        code = [
            'from gymnasium.wrappers import RecordVideo',
            'import os',
            'from IPython.display import Video',
            '',
            `env = gym.make("${envName}", render_mode='rgb_array')`,
            'video_folder = "./videos"',
            'os.makedirs(video_folder, exist_ok=True)',
            '',
            `env = RecordVideo(`,
            '    env,',
            '    video_folder=video_folder,',
            `    name_prefix="${videoPrefix}",`,
            `    episode_trigger=lambda e: e < ${nEpisodes}`,
            ')',
            '',
            `print("🎬 Enregistrement de ${nEpisodes} épisode(s)...")`,
            '',
            'for episode in range(' + nEpisodes + '):',
            '    state, info = env.reset()',
            '    total_reward = 0',
            '    terminated = truncated = False',
            '    ',
            '    while not (terminated or truncated):',
            '        action = env.action_space.sample()',
            '        state, reward, terminated, truncated, info = env.step(action)',
            '        total_reward += reward',
            '    ',
            `    print(f"Épisode {episode+1}: {total_reward:.1f} points")`,
            '',
            'env.close()',
            '',
            'video_path = f"{video_folder}/${videoPrefix}-ep0.mp4"',
            'print(f"\\n✅ Vidéo enregistrée: {video_path}")',
            'Video(video_path, embed=True, width=640)'
        ];
    } else if (agentType === 'qtable') {
        code = [
            'from gymnasium.wrappers import RecordVideo',
            'import os',
            'import numpy as np',
            'from IPython.display import Video',
            '',
            `env = gym.make("${envName}", render_mode='rgb_array')`,
            'video_folder = "./videos"',
            'os.makedirs(video_folder, exist_ok=True)',
            '',
            `env = RecordVideo(`,
            '    env,',
            '    video_folder=video_folder,',
            `    name_prefix="${videoPrefix}",`,
            `    episode_trigger=lambda e: e < ${nEpisodes}`,
            ')',
            '',
            `print("🎬 Enregistrement de ${nEpisodes} épisode(s) avec l'agent Q-Learning...")`,
            '',
            'for episode in range(' + nEpisodes + '):',
            '    state, info = env.reset()',
            '    total_reward = 0',
            '    terminated = truncated = False',
            '    ',
            '    while not (terminated or truncated):',
            '        action = np.argmax(q_table[state])',
            '        state, reward, terminated, truncated, info = env.step(action)',
            '        total_reward += reward',
            '    ',
            `    print(f"Épisode {episode+1}: {'Succès!' if envName.includes('FrozenLake') else total_reward > 0}: {{'Succès!' if total_reward > 0 else 'Échec'}} (reward={{total_reward}})")`,
            '',
            'env.close()',
            '',
            'video_path = f"{video_folder}/${videoPrefix}-ep0.mp4"',
            'print(f"\\n✅ Vidéo enregistrée: {video_path}")',
            'Video(video_path, embed=True, width=640)'
        ];
    } else {
        // model
        const modelType = modelPath.split('_')[0].toUpperCase();
        code = [
            'from gymnasium.wrappers import RecordVideo',
            'from stable_baselines3 import ' + modelType,
            'from IPython.display import Video',
            '',
            `model = ${modelType}.load("${modelPath}")`,
            '',
            `env = gym.make("${envName}", render_mode='rgb_array')`,
            'video_folder = "./videos"',
            'os.makedirs(video_folder, exist_ok=True)',
            '',
            `env = RecordVideo(`,
            '    env,',
            '    video_folder=video_folder,',
            `    name_prefix="${videoPrefix}",`,
            `    episode_trigger=lambda e: e < ${nEpisodes}`,
            ')',
            '',
            `print("🎬 Enregistrement de ${nEpisodes} épisode(s) avec l'agent entraîné...")`,
            '',
            'for episode in range(' + nEpisodes + '):',
            '    state, info = env.reset()',
            '    total_reward = 0',
            '    terminated = truncated = False',
            '    ',
            '    while not (terminated or truncated):',
            '        action, _ = model.predict(state, deterministic=True)',
            '        state, reward, terminated, truncated, info = env.step(action)',
            '        total_reward += reward',
            '    ',
            `    print(f"Épisode {episode+1}: {total_reward:.1f} points")`,
            '',
            'env.close()',
            '',
            'video_path = f"{video_folder}/${videoPrefix}-ep0.mp4"',
            'print(f"\\n✅ Vidéo enregistrée: {video_path}")',
            'Video(video_path, embed=True, width=640)'
        ];
    }
    
    // Trouver le bon endroit pour insérer (avant "Conclusion" ou après "Livrables")
    let insertIdx = nb.cells.length;
    for (let i = 0; i < nb.cells.length; i++) {
        if (nb.cells[i].cell_type === 'markdown') {
            const source = nb.cells[i].source.join('');
            if (source.includes('Livrables') && source.includes('Conclusion')) {
                insertIdx = i;
                break;
            }
        }
    }
    
    const markdownCell = {
        cell_type: 'markdown',
        metadata: {},
        source: [
            '---\n',
            '\n',
            '## 🎬 Enregistrement vidéo\n',
            '\n',
            description + '\n'
        ]
    };
    
    const codeCell = {
        cell_type: 'code',
        execution_count: null,
        metadata: {},
        outputs: [],
        source: code
    };
    
    // Insérer les cellules
    nb.cells.splice(insertIdx, 0, markdownCell, codeCell);
    
    fs.writeFileSync(notebookPath, JSON.stringify(nb, null, 1));
    console.log(`✅ Cellules vidéo ajoutées à ${notebookPath}`);
}

// Assurer que le dossier videos existe
fs.mkdirSync('./videos', { recursive: true });

console.log('='.repeat(50));
console.log('🎬 Ajout des cellules vidéo aux notebooks...');
console.log('='.repeat(50));

// Exercice 1: CartPole random (3 épisodes)
addVideoCell('exercice_01.ipynb', {
    envName: 'CartPole-v1',
    nEpisodes: 3,
    agentType: 'random',
    videoPrefix: 'exercice_01_cartpole_random',
    description: "L'agent **aléatoire** choisit ses actions au hasard. Observez comme il tombe rapidement !"
});

// Exercice 2: FrozenLake Q-Learning (3 épisodes)
addVideoCell('exercice_02.ipynb', {
    envName: 'FrozenLake-v1',
    nEpisodes: 3,
    agentType: 'qtable',
    videoPrefix: 'exercice_02_frozenlake_qlearning',
    description: "L'agent **Q-Learning** utilise la politique apprise. Observez comment il évite les trous !"
});

// Exercice 3: CartPole DQN (1 épisode)
addVideoCell('exercice_03.ipynb', {
    envName: 'CartPole-v1',
    nEpisodes: 1,
    agentType: 'model',
    modelPath: 'dqn_cartpole',
    videoPrefix: 'exercice_03_cartpole_dqn',
    description: "L'agent **DQN** entraîné maintient le pendule debout bien plus longtemps !"
});

// Mission: LunarLander PPO (1 épisode)
addVideoCell('mission.ipynb', {
    envName: 'LunarLander-v3',
    nEpisodes: 1,
    agentType: 'model',
    modelPath: 'ppo_lunar_lander',
    videoPrefix: 'mission_lunarlander_ppo',
    description: "L'agent **PPO** atterrit le module Eagle-1 sur la lune ! 🎯"
});

console.log('\n' + '='.repeat(50));
console.log('✅ Terminé !');
console.log('='.repeat(50));
console.log('\nProchaine étape:');
console.log('  1. git add .');
console.log('  2. git commit -m "Ajout des vidéos"');
console.log('  3. git push');
console.log('  4. Sur le pod: git pull && node add_videos.js');

File: notebooks/add_videos.py
#!/usr/bin/env python3
"""
Script pour ajouter des cellules d'enregistrement vidéo aux notebooks.
Exécute ce script depuis le dossier notebooks/ après un git pull.
"""
import json
import os
import sys

def add_video_cell(notebook_path, env_name, n_episodes, agent_type="random", model_path=None, video_prefix=None):
    """Ajoute une cellule d'enregistrement vidéo à un notebook."""
    
    if video_prefix is None:
        video_prefix = f"video_{env_name.lower().replace('-', '_').replace(' ', '_')}"
    
    with open(notebook_path, 'r') as f:
        nb = json.load(f)
    
    # Construire le code selon le type d'agent
    if agent_type == "random":
        code = f'''from gymnasium.wrappers import RecordVideo
from IPython.display import Video

env = gym.make("{env_name}", render_mode='rgb_array')
video_folder = "./videos"
os.makedirs(video_folder, exist_ok=True)

env = RecordVideo(
    env,
    video_folder=video_folder,
    name_prefix="{video_prefix}",
    episode_trigger=lambda e: e < {n_episodes}
)

print("🎬 Enregistrement de {n_episodes} épisode(s)...")

for episode in range({n_episodes}):
    state, info = env.reset()
    total_reward = 0
    terminated = truncated = False
    
    while not (terminated or truncated):
        action = env.action_space.sample()
        state, reward, terminated, truncated, info = env.step(action)
        total_reward += reward
    
    print(f"Épisode {{episode+1}}: {{total_reward:.1f}} points")

env.close()

video_path = f"{{video_folder}}/{video_prefix}-ep0.mp4"
print(f"\\n✅ Vidéo enregistrée: {{video_path}}")
Video(video_path, embed=True, width=640)'''
    elif agent_type == "qtable":
        code = f'''from gymnasium.wrappers import RecordVideo
from IPython.display import Video
import numpy as np

env = gym.make("{env_name}", render_mode='rgb_array')
video_folder = "./videos"
os.makedirs(video_folder, exist_ok=True)

env = RecordVideo(
    env,
    video_folder=video_folder,
    name_prefix="{video_prefix}",
    episode_trigger=lambda e: e < {n_episodes}
)

print("🎬 Enregistrement de {n_episodes} épisode(s) avec l'agent Q-Learning...")

for episode in range({n_episodes}):
    state, info = env.reset()
    total_reward = 0
    terminated = truncated = False
    
    while not (terminated or truncated):
        action = np.argmax(q_table[state])  # Politique apprise
        state, reward, terminated, truncated, info = env.step(action)
        total_reward += reward
    
    print(f"Épisode {{episode+1}}: {{'Succès!' if total_reward > 0 else 'Échec'}} (reward={{total_reward}})")

env.close()

video_path = f"{{video_folder}}/{video_prefix}-ep0.mp4"
print(f"\\n✅ Vidéo enregistrée: {{video_path}}")
Video(video_path, embed=True, width=640)'''
    else:  # model
        code = f'''from gymnasium.wrappers import RecordVideo
from stable_baselines3 import {model_path.split('_')[0].upper()}
from IPython.display import Video

model = {model_path.split('_')[0].upper()}.load("{model_path}")

env = gym.make("{env_name}", render_mode='rgb_array')
video_folder = "./videos"
os.makedirs(video_folder, exist_ok=True)

env = RecordVideo(
    env,
    video_folder=video_folder,
    name_prefix="{video_prefix}",
    episode_trigger=lambda e: e < {n_episodes}
)

print("🎬 Enregistrement de {n_episodes} épisode(s) avec l'agent entraîné...")

for episode in range({n_episodes}):
    state, info = env.reset()
    total_reward = 0
    terminated = truncated = False
    
    while not (terminated or truncated):
        action, _ = model.predict(state, deterministic=True)
        state, reward, terminated, truncated, info = env.step(action)
        total_reward += reward
    
    print(f"Épisode {{episode+1}}: {{total_reward:.1f}} points")

env.close()

video_path = f"{{video_folder}}/{video_prefix}-ep0.mp4"
print(f"\\n✅ Vidéo enregistrée: {{video_path}}")
Video(video_path, embed=True, width=640)'''
    
    # Trouver le bon endroit pour insérer (avant "Conclusion" ou fin du notebook)
    insert_idx = len(nb['cells'])
    for i, cell in enumerate(nb['cells']):
        if cell['cell_type'] == 'markdown':
            source = ''.join(cell.get('source', []))
            if 'Conclusion' in source and 'Livrables' in source:
                insert_idx = i
                break
    
    markdown_cell = {
        "cell_type": "markdown",
        "metadata": {},
        "source": [
            "---\n",
            "\n",
            "## 🎬 Enregistrement vidéo\n",
            "\n",
            f"Regardons l'agent {'aléatoire' if agent_type == 'random' else 'Q-Learning' if agent_type == 'qtable' else 'entraîné'} jouer en temps réel.\n"
        ]
    }
    
    code_cell = {
        "cell_type": "code",
        "execution_count": None,
        "metadata": {},
        "outputs": [],
        "source": [code]
    }
    
    # Insérer les cellules
    nb['cells'].insert(insert_idx, markdown_cell)
    nb['cells'].insert(insert_idx + 1, code_cell)
    
    with open(notebook_path, 'w') as f:
        json.dump(nb, f, indent=1)
    
    print(f"✅ Vidéo ajoutée à {notebook_path}")

def main():
    os.makedirs("./videos", exist_ok=True)
    
    print("=" * 50)
    print("Ajout des cellules vidéo aux notebooks...")
    print("=" * 50)
    
    # Exercice 1: CartPole random (3 épisodes)
    add_video_cell(
        "exercice_01.ipynb",
        env_name="CartPole-v1",
        n_episodes=3,
        agent_type="random",
        video_prefix="exercice_01_cartpole_random"
    )
    
    # Exercice 2: FrozenLake Q-Learning (3 épisodes)
    add_video_cell(
        "exercice_02.ipynb",
        env_name="FrozenLake-v1",
        n_episodes=3,
        agent_type="qtable",
        video_prefix="exercice_02_frozenlake_qlearning"
    )
    
    # Exercice 3: CartPole DQN (1 épisode)
    add_video_cell(
        "exercice_03.ipynb",
        env_name="CartPole-v1",
        n_episodes=1,
        agent_type="model",
        model_path="dqn_cartpole",
        video_prefix="exercice_03_cartpole_dqn"
    )
    
    # Mission: LunarLander PPO (1 épisode)
    add_video_cell(
        "mission.ipynb",
        env_name="LunarLander-v3",
        n_episodes=1,
        agent_type="model",
        model_path="ppo_lunar_lander",
        video_prefix="mission_lunarlander_ppo"
    )
    
    print("\n" + "=" * 50)
    print("Terminé ! Commit et push, puis git pull sur le pod.")
    print("=" * 50)

if __name__ == "__main__":
    main()
